docker/lr-pb/shard_bam.py

```diff
@@ -0,0 +1,124 @@
+    parser.add_argument('-p', '--prefix', type=str, default="shard", help="Shard filename prefix")
+    parser.add_argument('-n', '--num_shards', type=int, default=4, help="Number of shards")
+    parser.add_argument('-t', '--num_threads', type=int, default=2, help="Number of threads to use during sharding")
+    parser.add_argument('-i', '--index', type=str, required=False, help="Index filename")
+    parser.add_argument('bam', type=str, help="BAM")
+    args = parser.parse_args()
+
+    pbi = args.bam + ".pbi" if args.index is None else args.index
+
+    print(f"Reading index ({pbi}). This may take a few minutes...", flush=True)
+
+    # Decode PacBio .pbi file.  This is not a full decode of the index, only the parts we need for sharding.
+    # More on index format at https://pacbiofileformats.readthedocs.io/en/9.0/PacBioBamIndex.html .
+
+    fmt = Struct(
+        # Header
+        "magic" / Const(b"PBI\x01"),
+        "version_patch" / Int8ul,
+        "version_minor" / Int8ul,
+        "version_major" / Int8ul,
+        "version_empty" / Int8ul,
+        "pbi_flags" / Int16ul,
+        "n_reads" / Int32ul,
+        "reserved" / Padding(18),
+
+        # Basic information section (columnar format)
+        # "rgId" / LazyArray(this.n_reads, Int32sl),
+        # "qStart" / LazyArray(this.n_reads, Int32sl),
+        # "qEnd" / LazyArray(this.n_reads, Int32sl),
+        # "holeNumber" / Array(this.n_reads, Int32sl),
+        # "readQual" / LazyArray(this.n_reads, Float32b),
+        # "ctxtFlag" / LazyArray(this.n_reads, Int8ul),
+        # "fileOffset" / Array(this.n_reads, Int64sl),
+
+        "rgId" / Padding(this.n_reads * 4),
+        "qStart" / Padding(this.n_reads * 4),
+        "qEnd" / Padding(this.n_reads * 4),
+        "holeNumber" / Array(this.n_reads, Int32sl),
+        "readQual" / Padding(this.n_reads * 4),
+        "ctxtFlag" / Padding(this.n_reads * 1),
+        "fileOffset" / Array(this.n_reads, Int64sl),
+    )
+
+    # Make a list of bgzf virtual file offsets for sharding.
+    file_offsets_hash = OrderedDict()
+
+    last_offset = 0
+    with gzip.open(pbi, "rb") as f:
+        a = fmt.parse_stream(f)
+
+        for i in range(0, a.n_reads):
+            # Save only the virtual file offset for the first ZMW hole number, so that shard boundaries
+            # always keep reads from the same ZMW together.
+            if a.holeNumber[i] not in file_offsets_hash:
+                file_offsets_hash[a.holeNumber[i]] = a.fileOffset[i]
+
+            last_offset = a.fileOffset[i]
+
+    file_offsets = list(file_offsets_hash.values())
+
+    shard_offsets = []
+    for i in range(0, len(file_offsets), ceil(len(file_offsets)/args.num_shards)):
+        shard_offsets.append(file_offsets[i])
+
+    # For the last read in the file, pad the offset so the final comparison in write_shard() retains the final read.
+    shard_offsets.append(last_offset + 100)
```

Comment:
avoid magic numbers like 100. Maybe define it in a "macro" with an appropriate name

Reply:
Will fix.